src/main/java/com/google/devtools/build/lib/analysis/actions/PathMappers.java

```diff
@@ -36,6 +38,50 @@
  * PathMapper}).
  */
 public final class PathMappers {
+  private static final PathFragment BAZEL_OUT = PathFragment.create("bazel-out");
+  private static final PathFragment BLAZE_OUT = PathFragment.create("blaze-out");
+
+  /**
+   * A special instance for use in {@link AbstractAction#computeKey} when path mapping is generally
+   * enabled for an action.
+   *
+   * <p>When computing an action key, the following approaches to taking path mapping into account
+   * do <b>not</b> work:
+   *
+   * <ul>
+   *   <li>Using the actual path mapper is prohibitive since constructing it requires checking for
+   *       collisions among the action input's paths when computing the action key, which flattens
+   *       the input depsets of all actions that opt into path mapping and also increases CPU usage.
+   *   <li>Unconditionally using {@link StrippingPathMapper} can result in stale action keys when an
+   *       action is opted out of path mapping at execution time due to input path collisions after
+   *       stripping. See path_mapping_test for an example.
+   *   <li>Using {@link PathMapper#NOOP} does not distinguish between map_each results built from
```

Comment:
We can't get rid of the difference in this situation since `File#path` only returns the correct result when used in a `map_each` callback. If someone e.g. keeps paths as strings around in providers and then migrates to artifacts to become compatible with path mapping, they may run into this staleness (without this fix).